sdk/java/pulumi/src/main/java/com/pulumi/resources/internal/Stack.java

```diff
@@ -0,0 +1,109 @@
+
+import com.pulumi.core.Output;
+import com.pulumi.core.internal.annotations.InternalUse;
+import com.pulumi.deployment.internal.DeploymentInternal;
+import com.pulumi.resources.ComponentResource;
+import com.pulumi.resources.ComponentResourceOptions;
+import com.pulumi.resources.ResourceTransformation;
+
+import javax.annotation.ParametersAreNonnullByDefault;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+import static java.util.Objects.requireNonNull;
+
+@InternalUse
+@ParametersAreNonnullByDefault
+public final class Stack extends ComponentResource {
+
+    /**
+     * The type name that should be used to construct the root component in the tree of Pulumi resources
+     * allocated by a deployment. This must be kept up to date with
+     * "github.com/pulumi/pulumi/sdk/v3/go/common/resource/stack.RootStackType".
+     */
+    @InternalUse
+    public static final String RootPulumiStackTypeName = "pulumi:pulumi:Stack";
+
+    /**
+     * The outputs of this stack, if the <code>callback</code> callback exited normally.
+     */
+    private final Output<Map<String, Output<?>>> outputs;
+
+    /**
+     * Create a Stack with stack outputs defined in derived class constructor.
+     *
+     * @param projectName             the current project name
+     * @param stackName               the current stack name
+     * @param resourceTransformations optional list of transformations to apply to this stack's resources during construction.
+     *                                The transformations are applied in order, and are applied after all the transformations of custom
+     *                                and component resources in the stack.
+     * @param outputs                 the stack outputs
+     */
+    private Stack(
+            String projectName,
+            String stackName,
+            List<ResourceTransformation> resourceTransformations,
+            Output<Map<String, Output<?>>> outputs
+    ) {
+        super(
+                RootPulumiStackTypeName,
+                rootPulumiStackName(projectName, stackName),
+                ComponentResourceOptions.builder().resourceTransformations(
+                        resourceTransformations
+                ).build()
+        );
+        this.outputs = requireNonNull(outputs);
+    }
+
+    private static String rootPulumiStackName(String projectName, String stackName) {
+        return String.format("%s-%s", requireNonNull(projectName), requireNonNull(stackName));
+    }
+
+    @InternalUse
+    public Output<Map<String, Output<?>>> getOutputs() {
+        return this.outputs;
+    }
+
+    @InternalUse
+    public static Function<Supplier<Map<String, Output<?>>>, Stack> factory(
+            String projectName,
+            String stackName,
+            List<ResourceTransformation> resourceTransformations
+    ) {
```

Comment:
Wanted to make it clear what is the sequence of events and to have the whole init sequence in one place to help with maintaining and debugging in the future.